File: src/ObjectManager/Object.Tes/FilePacks/Records/045-TREE.Tree.cs
using OA.Core;

namespace OA.Tes.FilePacks.Records
{
    public class TREERecord : Record
    {
        public struct SNAMField
        {
            public int[] Values;

            public SNAMField(UnityBinaryReader r, int dataSize)
            {
                Values = new int[dataSize >> 2];
                for (var i = 0; i < Values.Length; i++)
                    Values[i] = r.ReadLEInt32();
            }
        }

        public struct CNAMField
        {
            public float LeafCurvature;
            public float MinimumLeafAngle;
            public float MaximumLeafAngle;
            public float BranchDimmingValue;
            public float LeafDimmingValue;
            public int ShadowRadius;
            public float RockSpeed;
            public float RustleSpeed;

            public CNAMField(UnityBinaryReader r, int dataSize)
            {
                LeafCurvature = r.ReadLESingle();
                MinimumLeafAngle = r.ReadLESingle();
                MaximumLeafAngle = r.ReadLESingle();
                BranchDimmingValue = r.ReadLESingle();
                LeafDimmingValue = r.ReadLESingle();
                ShadowRadius = r.ReadLEInt32();
                RockSpeed = r.ReadLESingle();
                RustleSpeed = r.ReadLESingle();
            }
        }

        public struct BNAMField
        {
            public float Width;
            public float Height;

            public BNAMField(UnityBinaryReader r, int dataSize)
            {
                Width = r.ReadLESingle();
                Height = r.ReadLESingle();
            }
        }

        public override string ToString() => $"TREE: {EDID.Value}";
        public STRVField EDID { get; set; } // Editor ID
        public MODLGroup MODL; // Model
        public FILEField ICON; // Leaf Texture
        public SNAMField SNAM; // SpeedTree Seeds, array of ints
        public CNAMField CNAM; // Tree Parameters
        public BNAMField BNAM; // Billboard Dimensions

        public override bool CreateField(UnityBinaryReader r, GameFormatId formatId, string type, int dataSize)
        {
            switch (type)
            {
                case "EDID": EDID = new STRVField(r, dataSize); return true;
                case "MODL": MODL = new MODLGroup(r, dataSize); return true;
                case "MODB": MODL.MODBField(r, dataSize); return true;
                case "MODT": MODL.MODTField(r, dataSize); return true;
                case "ICON": ICON = new FILEField(r, dataSize); return true;
                case "SNAM": SNAM = new SNAMField(r, dataSize); return true;
                case "CNAM": CNAM = new CNAMField(r, dataSize); return true;
                case "BNAM": BNAM = new BNAMField(r, dataSize); return true;
                default: return false;
            }
        }
    }
}

File: src.mac/ObjectManager/ObjectManager/Object.Tes/FilePacks/Records/045-TREE.Tree.swift
//
//  TREERecord.swift
//  ObjectManager
//
//  Created by Sky Morey on 5/28/18.
//  Copyright © 2018 Sky Morey. All rights reserved.
//

public class TREERecord: Record {
    public struct SNAMField {
        public var values: [Int32]

        init(_ r: BinaryReader, _ dataSize: Int) {
            values = [Int32](); let capacity = dataSize >> 2; values.reserveCapacity(capacity)
            for _ in 0..<capacity { values.append(r.readLEInt32()) }
        }
    }

    public struct CNAMField {
        public let leafCurvature: Float
        public let minimumLeafAngle: Float
        public let maximumLeafAngle: Float
        public let branchDimmingValue: Float
        public let leafDimmingValue: Float
        public let shadowRadius: Int32
        public let rockSpeed: Float
        public let rustleSpeed: Float

        init(_ r: BinaryReader, _ dataSize: Int) {
            leafCurvature = r.readLESingle()
            minimumLeafAngle = r.readLESingle()
            maximumLeafAngle = r.readLESingle()
            branchDimmingValue = r.readLESingle()
            leafDimmingValue = r.readLESingle()
            shadowRadius = r.readLEInt32()
            rockSpeed = r.readLESingle()
            rustleSpeed = r.readLESingle()
        }
    }

    public struct BNAMField {
        public let width: Float
        public let height: Float

        init(_ r: BinaryReader, _ dataSize: Int) {
            width = r.readLESingle()
            height = r.readLESingle()
        }
    }

    public override var description: String { return "TREE: \(EDID)" }
    public var EDID: STRVField = STRVField_empty // Editor ID
    public var MODL: MODLGroup? = nil // Model
    public var ICON: FILEField! // Leaf Texture
    public var SNAM: SNAMField! // SpeedTree Seeds, array of ints
    public var CNAM: CNAMField! // Tree Parameters
    public var BNAM: BNAMField! // Billboard Dimensions

    override func createField(_ r: BinaryReader, for format: GameFormatId, type: String, dataSize: Int) -> Bool {
        switch type {
        case "EDID": EDID = r.readSTRV(dataSize)
        case "MODL": MODL = MODLGroup(r, dataSize)
        case "MODB": MODL!.MODBField(r, dataSize)
        case "MODT": MODL!.MODTField(r, dataSize)
        case "ICON": ICON = r.readSTRV(dataSize)
        case "SNAM": SNAM = SNAMField(r, dataSize)
        case "CNAM": CNAM = CNAMField(r, dataSize)
        case "BNAM": BNAM = BNAMField(r, dataSize)
        default: return false
        }
        return true
    }
}
